vnext/Shared/RuntimeOptions.cpp

```diff
@@ -8,38 +8,104 @@ using std::mutex;
+  g_runtimeOptionInts.insert_or_assign(std::move(name), value ? 1 : 0);
 }
 
 void __cdecl SetRuntimeOptionInt(string &&name, int32_t value) noexcept {
   lock_guard<mutex> guard{g_runtimeOptionsMutex};
-  g_runtimeOptions.insert_or_assign(std::move(name), value);
+  g_runtimeOptionInts.insert_or_assign(std::move(name), value);
 }
 
 const bool __cdecl GetRuntimeOptionBool(const string &name) noexcept {
   lock_guard<mutex> guard{g_runtimeOptionsMutex};
-  auto itr = g_runtimeOptions.find(name);
-  if (itr != g_runtimeOptions.end())
+  auto itr = g_runtimeOptionInts.find(name);
+  if (itr != g_runtimeOptionInts.end())
     return itr->second == 1;
 
   return false;
 }
 
 const int32_t __cdecl GetRuntimeOptionInt(const string &name) noexcept {
   lock_guard<mutex> guard{g_runtimeOptionsMutex};
-  auto itr = g_runtimeOptions.find(name);
-  if (itr != g_runtimeOptions.end())
+  auto itr = g_runtimeOptionInts.find(name);
+  if (itr != g_runtimeOptionInts.end())
     return itr->second;
 
   return 0;
 }
 
 } // namespace Microsoft::React
+
+void __cdecl Microsoft_React_SetRuntimeOptionBool(const char *name, bool value) noexcept {
+  if (!name)
+    return;
+
+  lock_guard<mutex> guard{g_runtimeOptionsMutex};
+  if (value)
+    g_runtimeOptionInts.insert_or_assign(name, value ? 1 : 0);
+  else
+    g_runtimeOptionInts.erase(name);
+}
+
+void __cdecl Microsoft_React_SetRuntimeOptionInt(const char *name, int32_t value) noexcept {
+  if (!name)
+    return;
+
+  lock_guard<mutex> guard{g_runtimeOptionsMutex};
+  if (value)
+    g_runtimeOptionInts.insert_or_assign(name, value);
+  else
+    g_runtimeOptionInts.erase(name);
+}
+
+void __cdecl Microsoft_React_SetRuntimeOptionString(const char *name, const char *value) noexcept {
+  if (!name)
+    return;
+
+  lock_guard<mutex> guard{g_runtimeOptionsMutex};
+  if (value)
+    g_runtimeOptionStrings.insert_or_assign(name, value);
+  else
+    g_runtimeOptionStrings.erase(name);
+}
+
+const bool __cdecl Microsoft_React_GetRuntimeOptionBool(const char *name) noexcept {
+  lock_guard<mutex> guard{g_runtimeOptionsMutex};
+  auto itr = g_runtimeOptionInts.find(name);
+  if (itr != g_runtimeOptionInts.end())
+    return itr->second == 1;
+
+  return false;
+}
+
+const int32_t __cdecl Microsoft_React_GetRuntimeOptionInt(const char *name) noexcept {
+  lock_guard<mutex> guard{g_runtimeOptionsMutex};
+  auto itr = g_runtimeOptionInts.find(name);
+  if (itr != g_runtimeOptionInts.end())
+    return itr->second;
+
+  return 0;
+}
+
+const char *__cdecl Microsoft_React_GetRuntimeOptionString(const char *name) noexcept {
+  lock_guard<mutex> guard{g_runtimeOptionsMutex};
+  auto itr = g_runtimeOptionStrings.find(name);
+  if (itr != g_runtimeOptionStrings.cend()) {
+    auto size = itr->second.size() * sizeof(char) + 1 /*NULL termination*/;
+    auto result = static_cast<char *>(malloc(size));
+    strncpy_s(result, size, itr->second.c_str(), itr->second.size());
+
+    return result;
```

Comment:
>result [](http://example.com/codeflow?start=11&length=6)

We cannot do it: for ABI safety any memory allocated in a DLL must be deleted in the same DLL.

Normally we would have a method with two parameters: pointer to a buffer and the buffer size. The method would be called twice: first to measure and then with the allocated buffer. But we cannot do it here because we do all the work under the lock.

May I propose and alternative API that I had implemented in [https://github.com/microsoft/v8-jsi/blob/439a09ea1f645a83393b870992a121b6e54947cf/src/public/js_native_ext_api.h#L139](https://github.com/microsoft/v8-jsi/blob/439a09ea1f645a83393b870992a121b6e54947cf/src/public/js_native_ext_api.h#L139)
If we translate it here it would look like this:
```
// A callback to return string synchronously
typedef void (*GetStringCallback)(const char* data, size_t size, void *state);

void Microsoft_React_GetRuntimeOptionString(const char *name, GetStringCallback getString, void* state);
```
where the state is any data which we would normally capture in a lambda.